File: app/src/main/java/com/md/ModeHandler.java
package com.md;

import java.util.Stack;

import android.app.Activity;

import com.md.modesetters.DeckChooseModeSetter;
import com.md.modesetters.ModeSetter;

public class ModeHandler {

	Stack<ModeSetter> modeStack = new Stack<ModeSetter>();
	private final Activity context;

	public ModeHandler(Activity context) {
		this.context = context;

	}

	public ModeSetter whoseOnTop()
	{
		// Must have two.
		if (!modeStack.empty()) {
			return modeStack.peek();
		} else {
			return null;
		}
	}
	
	public boolean goBack() {
		// Must have two.
		if (modeStack.size() > 1) {
			ModeSetter pop = modeStack.pop();
			modeStack.peek().setupMode(context);
			return true;
		}
		return false;
	}

	public void add(ModeSetter modeSetter) {
		// Don't put your self on!
		if(modeStack.empty() || modeStack.peek() != modeSetter)
		{
		   modeStack.push(modeSetter);
		}
	}

}

File: app/src/main/java/com/md/ModeHandler.kt
package com.md

import android.app.Activity
import com.md.modesetters.ModeSetter
import com.md.modesetters.MoveManager.cancelJobs
import dagger.hilt.android.scopes.ActivityScoped
import java.util.Stack
import javax.inject.Inject

// TODOJ change to view model scoped.
@ActivityScoped
class ModeHandler @Inject constructor(val context: Activity) {
    private val modeStack = Stack<ModeSetter>()

    fun whoseOnTop(): ModeSetter? {
        // Must have two.
        return if (!modeStack.empty()) {
            modeStack.peek()
        } else {
            null
        }
    }

    fun goBack(): Boolean {
        // Must have two.
        if (modeStack.size > 1) {
            val pop = modeStack.pop()
            modeStack.peek().switchMode(context)
            return true
        }
        return false
    }

    fun add(modeSetter: ModeSetter) {
        cancelJobs()
        // Don't put your self on!
        if (modeStack.empty() || modeStack.peek() !== modeSetter) {
            modeStack.push(modeSetter)
        }
    }
}